README.md
# faceRecogDaemon

[![Python application](https://github.com/ohAnd/faceRecogDaemon/workflows/Python%20application/badge.svg)](https://github.com/ohAnd/faceRecogDaemon/actions?query=workflow%3A%22Python+application%22)
[![GitHub All Releases](https://img.shields.io/github/downloads/ohand/faceRecogDaemon/total)](https://github.com/ohAnd/faceRecogDaemon/releases)
[![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/ohand/faceRecogDaemon?include_prereleases)](https://github.com/ohAnd/faceRecogDaemon/releases)
[![GitHub issues](https://img.shields.io/github/issues-raw/ohand/faceRecogDaemon)](https://github.com/ohAnd/faceRecogDaemon/issues)
[![GitHub watchers](https://img.shields.io/github/watchers/ohand/faceRecogDaemon)](https://github.com/ohAnd/faceRecogDaemon/watchers)
[![GitHub contributors](https://img.shields.io/github/contributors-anon/ohand/faceRecogDaemon)](https://github.com/ohAnd/faceRecogDaemon/graphs/contributors)
[![GitHub commit activity](https://img.shields.io/github/commit-activity/m/ohand/faceRecogDaemon)](https://github.com/ohAnd/faceRecogDaemon/commits/master)

- [faceRecogDaemon](#facerecogdaemon)
  - [description](#description)
    - [Problem](#problem)
    - [Possible solution](#possible-solution)
    - [Features](#features)
      - [user side](#user-side)
      - [technical side](#technical-side)
  - [installation](#installation)
    - [download & start](#download--start)
    - [configuration](#configuration)
  - [API](#api)
    - [face recognition](#face-recognition)
      - [/api/check](#apicheck)
      - [/api/check?stream=1](#apicheckstream1)
    - [training](#training)
      - [/api/training?getdata=true](#apitraininggetdatatrue)
      - [/api/training?retrain=true](#apitrainingretraintrue)
      - [/api/training?rename=true&image=mmmmm.jpg&name=Joe](#apitrainingrenametrueimagemmmmmjpgnamejoe)
      - [/api/get_train_image?person=name&filename=name01.png](#apigettrainimagepersonnamefilenamename01png)
      - [/api/get_train_image?person=name&number=0](#apigettrainimagepersonnamenumber0)
    - [recent recognitions](#recent-recognitions)
      - [/api/get_recent_image](#apigetrecentimage)
      - [/api/get_recent_image?json](#apigetrecentimagejson)
      - [/api/get_recent_image?json&limit=30](#apigetrecentimagejsonlimit30)
      - [/api/get_recent_image?number=0](#apigetrecentimagenumber0)
      - [/api/get_recent_image?filename=name](#apigetrecentimagefilenamename)
      - [/api/get_recent_image?number=0&original=true](#apigetrecentimagenumber0originaltrue)
    - [images](#images)
      - [/api/get_image?file=name](#apigetimagefilename)
  - [Links](#links)

## description
Just started as a little application for easy managing face recognition outputs.

### Problem

You have a video/ image stream of a camera or something else. And you want to analyse the images for known faces. And get needed infos for further triggering for e.g. a smart home system to open a door or similar things.

### Possible solution

face recognition daamon ;-)

### Features

#### user side
- bind with image streams or NEW with rtsp stream, see config
- overview of recent historic recognized faces
  - change seen faces to other persons for next training
  - change seen faces to new persons for next training (coming soon)
- manage training data
  - remove images from person folder
  - move images to other persons
- manage settings (editing - coming soon)
- show api information

screenshots of client:

![home](./docs/home.png "home")
![home menu](./docs/home_menu.png "home menu")
![live](./docs/live.png "live")
![history](./docs/history.png "history")
![history training](./docs/history_train.png "history training")
![training](./docs/training.png "training")
![training images](./docs/training_images.png "training images")
![settings](./docs/settings.png "settings")
![api](./docs/api.png "api")

#### technical side
- for the recognition itself the library of https://github.com/ageitgey/face_recognition is used 
- api for all stuff provided
- client in hmtl/ javascript full responsive and prepared for web app
- 3-tier architecture - file-db, backend/ middleware, client ( not fully at this time ;-) )
  - "file-db" - primitve folder driven training data for images + archive for recognized faces
  - backend/ middleware - serves the functions over an api
  - client - html/ javascript web client prepared for a web app


## installation
The application is written in Python. So you need an already installed environment as described here: https://www.python.org/about/gettingstarted/ - needed version >= 3.6

If you running an ubuntu maybe you need for compiling dlib (only one time)

        sudo apt install python3.7-dev


### download & start

Donwload 'face recognition daemon' application the last release at https://github.com/ohAnd/faceRecogDaemon/releases

Update pip to the latest

        python -m pip3 install --upgrade pip

during the installation [dlib](dlib.net) will be installed (need's some time for compiling ;-) )

intall needed modules

        pip3 install -r requirements.txt

for quick start:
- create a path for known faces outside the app directory e.g. /path/to/knownFaces and put the path to config see below
  - inside this path create a dirextory for each person you want to recognize
  - in this subfolders you have to put some images for training
  - images must have only **one** face showing, otherwise it it will be ignored by training
    ```
      Structure:
            </path/to/knownFaces>/
            |-- <person1>/
            |   |-- <somename1>.jpeg
            |   |-- <somename2>.png
            |   |-- ...
            |-- <person2>/
            |   |-- <somename1>.jpeg
            |   |-- <somename2>.jpeg
            |-- ...
    ```
- create a archive path outside the app directory e.g. /path/to/imageArchive and put the path to config see below

start the server (service installation coming soon)

        bash://<paht/to/your/download>> ./faecRecogDaemon.py

### configuration
create a config file in application root directory

        touch config.ini

with following content

        # config file for face recog daemon
        [webserver]
        Port: 5001

        [imageSource]
        urlToImageSrc = <your_path_to_image_src> # e.g. http://<urlTocam>:port/image.cgi
        // urlToImageSrc = <your_path_to_image_src> # e.g. rtsp://user:pass@<ip-address>:<port>/media.amp

        [faceRecognition]
        knnDistance = 0.50

        [directories]
        pathToKnownFaces = <your_path>
        pathToImageArchive = <your_path>

___
___
## API
___
### face recognition 

#### /api/check
trigger new face recognization with a fresh image from source<br> return: JSON

<details>
<summary>response</summary>
with response code 200, following json will be returned

```json
{
  "checkDetails": {
    "NumberOfFaces": 0, 
    "imageToCheck": "http://<source image>", 
    "linkToArchivedImage": ""
  }, 
  "checkResults": {
    "linkToResultImage": "http://<host of face recog daemon>/api/get_image?file=/noFaces/<archived image name>.jpg", 
    "personsFound": {}
  }, 
  "status": "no faces found"
}
```

</details>

#### /api/check?stream=1
get online face recognization, return the result image directly<br> return: image
___
### training

#### /api/training?getdata=true
get all training src data as summary in JSON

<details>
<summary>response</summary>
with response code 200, following json will be returned

```json
{
  "persons": {
    "<nameOfPerson1>": {
      "count": 26, 
      "imagelinks": [
        "<baseImage>.png", 
        ...
        "<baseImage_n_>.jpg", 
        "<baseImage_m_>.jpg"
      ]
    }, 
    "<nameOfPerson2>": {
      "count": 10, 
      "imagelinks": [
        "<baseImage>.png", 
        ...
        "<baseImage_n_>.jpg", 
        "<baseImage_m_>.jpg"
      ]
    }
  }
}

```

</details>

#### /api/training?retrain=true
retrigger the training of given images

#### /api/training?rename=true&image=mmmmm.jpg&name=Joe
move image from history to training data with given person name

#### /api/get_train_image?person=name&filename=name01.png
get trained image by person and filename

#### /api/get_train_image?person=name&number=0
get trained image by number
___
### recent recognitions
#### /api/get_recent_image
get most recent image

#### /api/get_recent_image?json
get an overview of recent images in JSON<br>(default limited to 10, if less images available it is shortend)

#### /api/get_recent_image?json&limit=30
get an overview of recent images in JSON with given limit of resopnse<br>e.g. 30 entries (hard limited to 1000)

#### /api/get_recent_image?number=0
get most recent image - higher number means older image

#### /api/get_recent_image?filename=name
get most recent image with new face recognition fpr filename

#### /api/get_recent_image?number=0&original=true
get most recent image without image manipulation (timestamp face frame/ naming)
___
### images
#### /api/get_image?file=name
get image file with name

___                


## Links
source for used libs:
- https://github.com/ageitgey/face_recognition
- https://github.com/ageitgey/face_recognition/blob/master/examples/face_recognition_svm.py
- https://face-recognition.readthedocs.io/en/latest/face_recognition.html

used images:
- https://www.freepik.com/premium-vector/biometric-person-identification-facial-recognition-concept-futuristic-low-polygonal-human-face_4847694.htm
- ...




CHANGELOG.md
# Changelog

## v1.0.2 (16/02/2020)
- [due to broken connections at rtsp stream - added a timer thread to stop the streaming after some time of unusing](https://github.com/ohAnd/faceRecogDaemon/commit/9635d7c32248dc0b77aab6a3ab9bc962c49817d1) - @ohAnd
- [face locations with explicit model](https://github.com/ohAnd/faceRecogDaemon/commit/18af8eb859748135b9d3cef7d772ebedd22a5d99) - @ohAnd
- [[feature] full screen for liveview](https://github.com/ohAnd/faceRecogDaemon/commit/f7dbc961fbab6eb935c0f0aeee4ca94111761b7a) - @ohAnd
- [add links to badges](https://github.com/ohAnd/faceRecogDaemon/commit/9268794f315146eff75c8c50f3265d63f8f22f22) - @ohAnd
- [fix due to linitng and added perfomance measuring - logging](https://github.com/ohAnd/faceRecogDaemon/commit/3d28c854e37b6fb32812f81b59fac52623a6a713) - @ohAnd
- [updated to python3 and added feature - rtsp connection as image src](https://github.com/ohAnd/faceRecogDaemon/commit/d5140ff827deb94c8d2bb8ff49f0e9d0ca789cae) - @ohAnd
- [update bagdes for quick info](https://github.com/ohAnd/faceRecogDaemon/commit/c725124f2b67df7457a6976029341a172907a540) - @ohAnd
- [added first test for CD/CI](https://github.com/ohAnd/faceRecogDaemon/commit/1d9d5f8b982957b141ab6d332a82d20f04d8cfad) - @ohAnd
- [some corrections for CD/CI](https://github.com/ohAnd/faceRecogDaemon/commit/7a61fafc462374948c149771639132673f60ca2d) - @ohAnd
- [added requirements.txt for needed modules](https://github.com/ohAnd/faceRecogDaemon/commit/8cedafce32d1ba773ec51618699124c678be5192) - @ohAnd


modules/test_imageHandling.py
import pytest
import imageHandling

def test_getImageFromUrl():
    assert imageHandling.getImageFromUrl('http://localhost:1234') == 'connection_timeout'

.grenrc.js
/* .grenrc.js */
module.exports = 
{
    "dataSource": "commits",
    "includeMessages": "merge",
    "ignore-commits-with": [
        "pythonapp.yml",
        "set theme",
        "Update issue templates",
        "version update"
    ],
    "ignoreIssuesWith": [
        "wontfix",
        "duplicate"
    ],
    "template": {
        "commit": ({ message, url, author, name }) => `- [${message}](${url}) - ${author ? `@${author}` : name}`,
        "issue": "- {{labels}} {{name}} [{{text}}]({{url}})",
        "label": "[**{{label}}**]",
        "noLabel": "closed",
        "group": "\n#### {{heading}}\n",
        "changelogTitle": "# Changelog\n\n",
        "release": "## {{release}} ({{date}})\n{{body}}",
        "releaseSeparator": "\n---\n\n"
    },
    "groupBy": {
        "Enhancements:": ["enhancement", "internal","feature"],
        "Bug Fixes:": ["bug","fix"],
        "Other": ["*"]
    }
}

faceRecogDaemon.py
#!/usr/bin/python3

portForWebserver = 5001
urlToImageSrc = 'https://c.pxhere.com/images/57/9c/893d3c322850a321e164a0c8cc75-1418580.jpg!d'

knnDistance = 0.6

pathToKnownFaces = 'train_dir'
pathToImageArchive = 'webserver2'

version = '1.0.2'

################################################################################################################################
import os
import sys
sys.path.append("modules/")
import imageHandling
import faceRecogKNN
import configparser

from flask import Flask, jsonify, request, send_file, render_template, g, \
     abort, Response, redirect, url_for
from datetime import datetime
app = Flask(__name__)

cfg = configparser.ConfigParser()
cfg.read("config.ini")
try:
    if cfg.get('webserver','Port') != '':
        port = cfg.get('webserver','Port')
    if cfg.get('imageSource','urlToImageSrc') != '':
        urlToImageSrc = cfg.get('imageSource','urlToImageSrc')
    if cfg.get('faceRecognition','knnDistance') != '':
        knnDistance = cfg.getfloat('faceRecognition','knnDistance')
        app.logger.info("config: "+str(knnDistance))
    if cfg.get('directories','pathToImageArchive') != '':
        pathToImageArchive = cfg.get('directories','pathToImageArchive')
    if cfg.get('directories','pathToKnownFaces') != '':
        pathToKnownFaces = cfg.get('directories','pathToKnownFaces')
except configparser.NoSectionError:
    app.logger.info("no config.ini - use defaults")
except configparser.NoOptionError:
    app.logger.info("config.ini - missing option")

# initialize rtsp instance for later threading, if rtsp url is given
if(urlToImageSrc.startswith("rtsp")):
    newRTSP = imageHandling.RTSPstream(urlToImageSrc)
global debug
debug = False
# # # # Webserver # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 

@app.route('/api/check', methods=['GET'])
def check():
    if debug: print (str(datetime.now().strftime("%Y%m%d_%H%M%S.%f"))+" - API-check 1 start gettinf rtsp frame")
    startProcessTime = datetime.now()
    # is online streaming requested
    if request.args.get('stream') == '1':
        stream = 1
    else:
        stream = 0
    checkResponse = {
        "status": 'not started',
        "checkDetails": {
            'NumberOfFaces': 0,
            'imageToCheck': '',
            'linkToArchivedImage': '',
            },
        "checkResults": {
            "personsFound": {},
            "linkToResultImage": ''
            }
    }
    # get last image
    
    if(urlToImageSrc.startswith("rtsp")):
        # start rtsp stream with first image request and start a timer to stop after 60 sec
        # TODO: restsart timer with every request to get the stream alive        
        if newRTSP.isNotRunning():
            newRTSP.start()
            timer = imageHandling.startTimer(newRTSP.stop)
            timer.start()
        # get image from rtsp stream
        lastImage = imageHandling.getImageFromRTSP(newRTSP)
                
        # send loading image if last image couldn't get (at starting this happnes sometimes)
        if lastImage is None:
            pathFilename = 'static/img/loading.png'
            return send_file(pathFilename, mimetype='image/gif')           
    else:
        lastImage = imageHandling.getImageFromUrl(urlToImageSrc)
    
    if "error" in lastImage:
        checkResponse['status'] = lastImage
        return jsonify(checkResponse)
    else:
        checkResponse['checkDetails']['imageToCheck'] = urlToImageSrc
    # get face Recog
    if debug: print (str(datetime.now().strftime("%Y%m%d_%H%M%S.%f"))+" -     1 start getFaces on frame")
    startProcessTimeFaces = datetime.now()
    noOfFaces = faceRecogKNN.getFacesInImage(lastImage)
    processTime = (datetime.now() - startProcessTimeFaces)
    if debug: print (str(datetime.now().strftime("%Y%m%d_%H%M%S.%f"))+" -     2 done  getFaces on frame: "+str(processTime))
    checkResponse['checkDetails']['NumberOfFaces'] = noOfFaces
    # get predictions for found faces
    personsFound = {} 
    imageBlob = ''
    if debug: print (str(datetime.now().strftime("%Y%m%d_%H%M%S.%f"))+" -     1 start prediction on frame")
    startProcessTimePred = datetime.now()
    predictions = faceRecogKNN.predict(lastImage, model_path=pathToKnownFaces+"/trained_knn_model.clf",distance_threshold=knnDistance)
    processTime = (datetime.now() - startProcessTimePred)
    if debug: print (str(datetime.now().strftime("%Y%m%d_%H%M%S.%f"))+" -     2 done  prediction on frame: "+str(processTime))
    if noOfFaces > 0:
        for name, (top, right, bottom, left) in predictions:
            if name in personsFound:
                name = name+'_'+str(datetime.now().strftime("%Y%m%d_%H%M%S"))
            personsFound[name] = {
                "left" : left,
                "top": top
            }
        checkResponse['status'] = 'ok'
        checkResponse['checkResults']['personsFound'] = personsFound
    else:
        checkResponse['status'] = 'no faces found'
    # save prediction image
    if stream == 0:
        imageLocalLink = faceRecogKNN.show_prediction_labels_on_image(lastImage,pathToImageArchive,predictions,0,noOfFaces)
        checkResponse['checkResults']['linkToResultImage'] = 'http://'+request.host +'/api/get_image?file='+imageLocalLink
    else:
        if debug: print (str(datetime.now().strftime("%Y%m%d_%H%M%S.%f"))+" -     1 start recog on frame")
        startProcessTimeRecog = datetime.now()
        imageBlob = faceRecogKNN.show_prediction_labels_on_image(lastImage,pathToImageArchive,predictions,0,noOfFaces,retLinkOrBlob='blob')
        processTime = (datetime.now() - startProcessTimeRecog)
        if debug: print (str(datetime.now().strftime("%Y%m%d_%H%M%S.%f"))+" -     2 done  recog on frame: "+str(processTime))
        
    if stream == 1:
        # print (str(datetime.now().strftime("%Y%m%d_%H%M%S.%f"))+" - 4 send image to client")
        processTime = (datetime.now() - startProcessTime)
        if debug: print (str(datetime.now().strftime("%Y%m%d_%H%M%S.%f"))+" - API-Check 2 done - send image to client: "+str(processTime))
        return send_file(imageBlob, mimetype='image/gif')
    else:
        # Return the result as json
        return jsonify(checkResponse)

@app.route('/api/retrain', methods=['GET'])
def reTrain():
    app.logger.info("Training KNN classifier...")
    classifier = faceRecogKNN.train(pathToKnownFaces, model_save_path=pathToKnownFaces+"/trained_knn_model.clf", n_neighbors=2)
    app.logger.info("Training complete!")
    # Return the result as json
    result = {
        "training": 'done',
    }
    return jsonify(result)

@app.route('/api/training', methods=['GET', 'POST'])
def apiTraining():
    if request.method == 'GET' and request.args.get('getdata') == 'true':
        persons = imageHandling.getPersonFromTrainFolder(pathToKnownFaces)
        result = {
            "persons": persons
        }
        return jsonify(result)
    elif request.method == 'GET' and request.args.get('retrain') == 'true':
        # import time
        def generate():
            # app.logger.info('request started')
            # for i in range(5):
            #     time.sleep(1)
            #     yield str(i)
            # app.logger.info('request finished')
            startProcessTime = datetime.now()
            app.logger.info("Training KNN classifier...")
            n_neighbors = faceRecogKNN.train(pathToKnownFaces, model_save_path=pathToKnownFaces+"/trained_knn_model.clf", verbose=True) #, n_neighbors=2
            app.logger.info("Training complete!")
            processTime = (datetime.now() - startProcessTime)
            processTime = str(processTime).split('.')[0]
            yield str(processTime + " with " + str(n_neighbors) + " neighbors")
        return Response(generate(), mimetype='text/plain')
    elif request.method == 'GET' and request.args.get('rename') == 'true':
        src = pathToImageArchive+request.args.get('image','')
        tgt = pathToKnownFaces+'/'+request.args.get('name','')+'/'+os.path.basename(request.args.get('image',''))
        taggedImageFile = imageHandling.getTaggedImageOfOriginal(src)
        taggedImage = pathToImageArchive+'/faces/tagged/'+taggedImageFile
        # def generate():
        app.logger.info("add image to train dir - src: "+src+" dst: "+tgt)
        resp = imageHandling.copyFileToFolder(src,tgt)
        resp = imageHandling.deleteFileWithPath(taggedImage)
            # if resp != 0:
            #     yield resp
            # yield ''
        # return 0 Response(generate(), mimetype='text/plain')
        return redirect(url_for('.history', _anchor=request.args.get('image','')))
    abort(404)

@app.route('/api/get_image')
def get_image():
    filename = request.args.get('file')
    pathFilename = os.path.abspath(pathToImageArchive+'/'+filename)
    if not os.path.isfile(pathFilename):
        pathFilename = 'static/img/404.png'
    return send_file(pathFilename, mimetype='image/gif')

@app.route('/api/get_recent_image')
def get_recent_image():
    if 'json' in request.args: # request.args.get('json','false') == 'true':
        limit = int(request.args.get('limit',10)) # limit the response to default 10 recent images
        if limit > 1000:
            limit = 1000
        images=[]
        for imgNo in range(0,limit):
            imgObj = {}
            apiLink     = "http://"+request.host+'/api/get_image?file='
            apiLinkNew  = "http://"+request.host+'/api/get_recent_image?filename='
            subFolderWithFaces = "/faces"            
            image = imageHandling.getRecentFileOfDirectory(pathToImageArchive+subFolderWithFaces,imgNo)
            if image == 'notFound':
                continue
            
            taggedImage = imageHandling.getTaggedImageOfOriginal(image)
            
            imageOriginalLink   = apiLinkNew+image.replace(pathToImageArchive,'')
            if taggedImage != 'notFound':
                imageTaggedLink = apiLink+subFolderWithFaces+'/tagged/'+ taggedImage
                taggedName = taggedImage.split('__')[1].split('.jpg')[0]
            else:
                imageTaggedLink = 'notFound'
                taggedName = 'notFound'
                            
            # tagged = apiLink+image.replace(pathToImageArchive,"")
            # orig = apiLink+image.replace(pathToImageArchive+'/faces/',"")
            # orig = orig.split('__')[0]+'.jpg'
            # origImageName = orig.split(apiLink)[1]
            
            imgObj['imageTaggedLink']   = imageTaggedLink
            imgObj['imageOriginalLink'] = imageOriginalLink
            imgObj['imageOriginalName'] = image.replace(pathToImageArchive,'')

            imgObj['name'] = taggedName
            images.append(imgObj)
        result = {
        "recentImages": images,
        "overallRecognitons": str(imageHandling.getRecentFileOfDirectoryMaxCount(pathToImageArchive+subFolderWithFaces))
        }
        return jsonify(result)
    elif 'filename' in request.args:
        pathFilename = pathToImageArchive+request.args.get('filename')
    elif 'number' in request.args:
        recentImageNumber = request.args.get('number')
        pathFilename = imageHandling.getRecentFileOfDirectory(pathToImageArchive+'/faces/tagged/',recentImageNumber,'jpg')
        return send_file(pathFilename, mimetype='image/gif')
    else: # no args given, most recent image
        pathFilename = imageHandling.getRecentFileOfDirectory(pathToImageArchive+'/faces/tagged/',0)
    app.logger.info("/api/get_recent_image - given filename: "+pathFilename)
    # compose the response
    if not os.path.isfile(pathFilename):
        pathFilename = 'static/img/404.png'
        return send_file(pathFilename, mimetype='image/gif') 
    elif request.args.get('original','false') == 'true':
        if not os.path.isfile(pathFilename):
            pathFilename = 'static/img/404.png'
        return send_file(pathFilename, mimetype='image/gif') 
    else:
        startProcessTime = datetime.now()
        lastImage = pathFilename
        timeStampOfFile = imageHandling.getCtimeFormmatedOfFile(lastImage)
        predictions = faceRecogKNN.predict(lastImage, model_path=pathToKnownFaces+"/trained_knn_model.clf",distance_threshold=knnDistance)
        imageBlob = faceRecogKNN.show_prediction_labels_on_image(lastImage,pathToImageArchive,predictions,timeStampOfFile,retLinkOrBlob='blob')
        processTime = (datetime.now() - startProcessTime)
        app.logger.info("recogonition for: "+lastImage+" needs: "+str(processTime))
        return send_file(imageBlob, mimetype='image/gif')

@app.route('/api/get_train_image')
def get_train_image():
    imageToShow = request.args.get('number',0) # if not set take most recent -> 0
    person = request.args.get('person','')
    if 'filename' in request.args: 
        pathFilename = pathToKnownFaces+"/"+person+"/"+request.args.get('filename',0)
    else:
        pathFilename = imageHandling.getImageInTrainDirectory(pathToKnownFaces,person,imageToShow)
    #print "path: "+pathFilename
    if not os.path.isfile(pathFilename):
        pathFilename = 'static/img/404.png'
    return send_file(pathFilename, mimetype='image/gif')

# website #########################################################################

@app.route('/')
def index():
    g.version = version
    # imageHandling.getFileProps('eee.jpgneu1.jpg')
    return  render_template("index.html")

@app.route('/liveview')
def liveview():
    values = {}
    values['hostlink'] = request.host
    g.version = version
    return  render_template("liveview.html", values=values)

@app.route('/history')
def history():
    values = {}
    values['hostlink'] = request.host
    g.version = version
    return  render_template("history.html", values=values)

@app.route('/training')
def training():
    g.version = version
    values = {}
    values['hostlink'] = request.host
    persons = imageHandling.getPersonFromTrainFolder(pathToKnownFaces)
    return  render_template("training.html", persons=persons, values=values)

@app.route('/settings')
def settings():
    config = {}
    config['urlToImageSrc'] = urlToImageSrc
    config['pathToImageArchive'] = pathToImageArchive
    config['knnDistance'] = knnDistance
    g.version = version
    return  render_template("settings.html", config=config)

@app.route('/apiinfo')
def apiinfo():
    values = {}
    values['hostlink'] = request.host
    g.version = version
    return  render_template("apiinfo.html", values=values)

if __name__ == "__main__":
    #faceRecog.init_faceRecog(pathToKnownFaces)
    if not os.path.isfile(pathToKnownFaces+"/trained_knn_model.clf"):
        app.logger.info("Training KNN classifier...")
        classifier = faceRecogKNN.train(pathToKnownFaces, model_save_path=pathToKnownFaces+"/trained_knn_model.clf", n_neighbors=2)
        app.logger.info("Training complete!")

    import logging
    logging.basicConfig(level=logging.INFO,filename=pathToImageArchive+"/faceRecogDeamon.log")
    logger = logging.getLogger(__name__)

    # httpLogger = logging.getLogger("*")
    # httpLogger.addHandler(logging.FileHandler("http-log.txt"))
    # # httpLogger.addFilter(logging.Filter("HTTP"))
    # httpLogger.propagate = False
    
    # app.run(host='0.0.0.0', port=portForWebserver, debug=True)
    from gevent.pywsgi import WSGIServer
    http_server = WSGIServer(('', portForWebserver), app, log=logger)
    http_server.serve_forever()